src/Markup/Avalonia.Markup.Xaml/SourceInfo/SourceInfo.cs

```diff
@@ -0,0 +1,215 @@
+    /// source location information (file path, line, and column) with UI elements.
+    /// <para/>
+    /// The <see cref="SourceInfo"/> property is typically populated automatically by the Avalonia XAML compiler:
+    /// <list type="number">
+    /// <item>
+    /// When running in <b>design mode</b> — to enable designer tools to map rendered elements back to source XAML.
+    /// </item>
+    /// <item>
+    /// When running in a <b>debug configuration</b> — allowing runtime inspection or navigation
+    /// back to the originating XAML source (this can be overridden by setting the 
+    /// <c>AvaloniaXamlCreateSourceInfo</c> build property).
+    /// </item>
+    /// </list>
+    /// </remarks>
+    public static class Source
+    {
+        /// <summary>
+        /// Defines the attached <see cref="SourceInfo"/> property that stores the source location
+        /// information for a control.
+        /// </summary>
+        public static readonly AttachedProperty<SourceInfo> SourceInfoProperty =
+            AvaloniaProperty.RegisterAttached<AvaloniaObject, SourceInfo>(
+                "SourceInfo",
+                typeof(Source));
+
+        /// <summary>
+        /// Gets the <see cref="SourceInfo"/> value associated with the specified visual.
+        /// </summary>
+        /// <param name="v">The control from which to retrieve the source information.</param>
+        /// <returns>
+        /// The <see cref="SourceInfo"/> that describes where the control was defined in XAML.
+        /// If no value is set, the default (empty) <see cref="SourceInfo"/> is returned.
+        /// </returns>
+        public static SourceInfo GetSourceInfo(AvaloniaObject v)
+        {
+            return v.GetValue(SourceInfoProperty);
+        }
+
+        /// <summary>
+        /// Sets the <see cref="SourceInfo"/> value for the specified control.
+        /// </summary>
+        /// <param name="v">The control to associate with the source information.</param>
+        /// <param name="value">The <see cref="SourceInfo"/> describing the control’s origin in XAML.</param>
+        /// <remarks>
+        /// Normally this method is invoked automatically by the XAML compiler or design-time tools.
+        /// You generally do not need to set this property manually unless you are generating controls dynamically.
+        /// </remarks>
+        public static void SetSourceInfo(AvaloniaObject v, SourceInfo value)
+        {
+            v.SetValue(SourceInfoProperty, value);
+        }
+    }
+
+    /// <summary>
+    /// Attribute applied to types generated from XAML to record the original source file name.
+    /// </summary>
+    /// <remarks>
+    /// The XAML compiler may add this attribute to generated classes so design-time tooling
+    /// can map a runtime type back to its originating XAML file.
+    /// </remarks>
+    [AttributeUsage(AttributeTargets.Class, Inherited = false)]
+    public sealed class XamlSourceInfoAttribute : Attribute
```

Comment:
I assume this is needed because:

```
// Add [XamlSourceInfoAttribute] to the generated class.
// Used at design time to locate the original .axaml file,
// since the runtime loader substitutes a fake file name.
```

This feels like a bit of a hack. Is it not possible to make the runtime loader use the correct filename?